Ch 5 edits

carols10cents opened the pull request:
I'm not done with chapter 5 yet, but I got through the first section and wanted to see if yinz totally hate this direction or not.

The goal of these changes is to remove concepts that distract from structs, namely the namespacing of the `f32::powi` functions, the difference between `powi` and `powf`, and the `Copy` and `Clone` traits.

I went with borrowing instead of showing `Copy` to better set up methods-- the first thing the methods section will now show is changing the `area` function to be a method that takes `&self`, which I think will clear up that part.

Then we'll add another method to Rectangle that takes an argument-- I'm thinking about computing whether another rectangle intersects with this one? (which means rectangles will have to have x,y coords as well) or whether this rectangle could encompass the other rectangle completely... something along those lines. This would *only* introduce `other`, so that it's not introduced at basically the same time as `self` and maybe will be less confusing.

Soooo do you hate it y/n?

steveklabnik commented on src/ch05-00-structs.md:
```diff
@@ -1,301 +1,271 @@
-from an object-oriented language, a `struct` is like an object's data
-attributes. `struct` and `enum` (that we will talk about in the next chapter)
-are the building blocks you can use in Rust to create new types in your
+A `struct`, short for "structure", is a custom data type that lets us name and
```
should be single quotes

steveklabnik commented on src/ch05-00-structs.md:
````diff
@@ -1,301 +1,271 @@
 ```rust,ignore
-f64::powi(2.0, 3)
+let user1 = User {
+    email: "someone@example.com",
````
this isn't a `String`

carols10cents replied:
efff

steveklabnik commented on src/ch05-00-structs.md:
````diff
@@ -1,301 +1,271 @@
-    f64::sqrt(x_squared + y_squared)
-}
+```bash
+The area of the rectangle is 1500
````
are nerds gonna Get Mad because this has no units?

carols10cents replied:
probably... is adding them in the println enough?

steveklabnik replied:
yeah, sounds good

steveklabnik commented on src/ch05-00-structs.md:
```diff
@@ -1,301 +1,271 @@
-this:
-
-Filename: src/main.rs
+If we run this, we get an error with the core message ``the trait bound
```
double graves here are gonna fail to render properly

carols10cents replied:
:( :( :( i'll figure out how to actually do it

carols10cents replied:
ugh that is the commonmark way to do it, i guess pulldown-cmark doesn't conform. i don't see any bugs/commits/tests about this, though, maybe i'll open one later. I'll just take out the backticks in the message for now.

carols10cents replied:
actually it looks crappy that way too, code block it is

steveklabnik replied:
oh, well if it renders okay, that's the thing that matters

carols10cents replied:
it does not

steveklabnik commented on src/ch05-00-structs.md:
```diff
@@ -1,301 +1,271 @@
+Because of this ambiguity, Rust doesn't try to guess what we want and structs
+do not have a provided implementation of `Display`.
+
+If we keep reading the error messages, though, we'll find ``note: `Rectangle`
```
same here

carols10cents commented:
Ok, since you don't seem to hate this, I'm going to continue on with this example through Methods :) I'm fine just leaving this PR open until I've got the whole chapter up; thank you for taking a look along the way! ❤️

carols10cents commented:
Ok, I think I'm done. Well, maybe. I took out the part about automatic referencing-- I haven't really seen anyone notice that or have problems with it. If it was in there to explain to C++ devs why there's only `.` and not `->`, maybe we could make it be in a box and be a bit off to the side?

I also added a bit about associated functions, I feel like it fits in here and I was a little surprised it wasn't there. Maybe I moved stuff around too much and lost it somewhere.

steveklabnik commented:
> If it was in there to explain to C++ devs why there's only . and not ->, maybe we could make it be in a box and be a bit off to the side?

:+1:

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -1,189 +1,170 @@
+their name, they can take arguments and return values, and they contain some
+code that gets run when they're called from somewhere else. Methods are
+different from functions, however, because they're defined within the context
+of a struct (or an enum or a trait object, but we're going to focus on structs
```
we haven't come close to talking about trait objects yet, maybe just say "other things"

carols10cents replied:
mmm liz doesn't seem to like "things" very much. I don't think we've talked about enums yet either. What if I change this to "Or an enum, which we'll cover in Chapter 6, or a trait object, which we'll cover in Chapter XX"?

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -1,189 +1,170 @@
 
-When we defined `distance()`, we took both `self` and the other argument by
-reference. Yet, we needed a `&` for `p2` but not `p1`. What gives?
+One more useful feature of `impl` blocks: we're allowed to define functions within `impl` blocks that *don't* take `self` as a parameter. These are called *associated functions*, since they're associated with the struct. They're still functions though, not methods, since they don't have an instance of the struct to work with. You've already used an associated function: `String::from`.
```
wrap to 80 cols

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -1,189 +1,170 @@
-of the few places in Rust that has behavior like this. Here’s how it works:
-when you call a method with `self.(`, Rust will automatically add in `&`s
-or `&mut`s to match the signature. In other words, these are the same:
+Associated functions are often used for constructors that will return a new instance of the struct. For example, we could provide an associated function that would take one dimension argument and use that as both length and width, thus making it easier to create a square `Rectangle` rather than having to specify the same value twice:
```
same

carols10cents commented:
You should definitely read over this commit and make sure I'm not lying.... I think I might have said some things backwards maybe... https://github.com/rust-lang/book/pull/266/commits/6947ef48cc15caebdffef7071c823244a66e0675

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -69,6 +69,51 @@ signature, is for organization. We've put all the things we can do with an
+
+#### Where's the `->` operator?
+
+In C++, there are two different operators for calling methods: `.` if you're
```
Maybe "in languages like C or C++"

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -69,6 +69,51 @@ signature, is for organization. We've put all the things we can do with an
+#### Where's the `->` operator?
+
+In C++, there are two different operators for calling methods: `.` if you're
+calling a method on the object directly, and `->` if you're calling the method
```
though, given that you're talking about _methods_ here... C doesn't have methods.

carols10cents replied:
yeah i thought about saying "fields and methods" but this is the methods chapter but i guess this is also the structs chapter but... let me try

carols10cents replied:
i think "languages like c++" invites the reader to extrapolate to C in the ways that C is like C++, so I think that'll work

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -69,6 +69,51 @@ signature, is for organization. We've put all the things we can do with an
+
+In C++, there are two different operators for calling methods: `.` if you're
+calling a method on the object directly, and `->` if you're calling the method
+on a pointer to the object and thus need to dereference the pointer first. Rust
```
I wonder if it's worth saying something like "In other words, if `foo` is a pointer, `foo->bar()` is like `(*foo).bar()`.

steveklabnik commented on src/ch05-01-method-syntax.md:
```diff
@@ -69,6 +69,51 @@ signature, is for organization. We've put all the things we can do with an
+Rust that has behavior like this.
+
+Here’s how it works: when you call a method with `object.something(`, Rust will
+automatically add in `&`s or `&mut`s so that `object` matches the signature of
```
I _think_ it will try `*`s too?

carols10cents replied:
that's auto *de*referencing tho, right? i think we should say that here though too because that's the part that actually matches the C++, which is what i was thinking when i said i was pretty sure i was saying stuff backwards. I think.

carols10cents commented:
Tag you're it!

steveklabnik commented:
I'm feeling good. Shall we merge?

carols10cents commented:
Yayayayayaya!!!
